<a href="https://colab.research.google.com/github/dipesh2108/AI_Notes/blob/main/8_Random_Forest__ensemble_method.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Random Forest Algorithm with Python and Scikit-Learn
-------------------------------------------------------------------------------

Random forest is a type of supervised machine learning algorithm based on **ensemble learning**. Ensemble learning is a type of learning where you join different types of algorithms or same algorithm multiple times to form a more powerful prediction model. The random forest algorithm combines multiple algorithm of the same type i.e. multiple decision trees, resulting in a forest of trees, hence the name "Random Forest". The random forest algorithm can be used for both regression and classification tasks.

How the Random Forest Algorithm Works
------------------------------------------------------------

The following are the basic steps involved in performing the random forest algorithm:

1> Pick N random records from the dataset, **call it Bootstraped Dataset**. <br>
2> Build a decision tree based on these N records. <br>
3> Choose the number of trees you want in your algorithm
   and repeat steps 1 and 2. <br>
4> In case of a **regression problem**, <font color='green'> for a new record, each tree in the forest predicts a value for Y (output). The final value can be calculated by taking the average of all the values predicted by all the trees in forest.</font> <br>
Or, in case of a **classification problem**, <font color='green'> each tree in the forest predicts the category to which the new record belongs. Finally, the new record is assigned to the category that wins the majority vote. </font>

<h3>Quick watch this <b>How does Random Forest work ?</b></h3>

<a href="https://drive.google.com/open?id=1FYcZj1CggyWiAkq0JhoyqZh9-n3mniGk">
  <img src="https://drive.google.com/uc?id=14OOsd0HaKoMJjqu5YT5n7-HsvE6UVV7z" alt="How Random Forest work ?" width="130" height="70">
</a>

<small><b>This Video would give you a fare idea on <u>What will we</u> do in code ahead </b><small><br>
<small>Credits : This video is recorded by Josh Starmer for educational purposes.</small>

<hr>

**`Question time : (see the video again to answer)`**

Q. <font color='red'><b> What does Bagging mean ?</b></font>

Q. <font color='red'><b> Is their something like "Out of Bag" Error ?</b></font>
<hr>

Advantages of using Random Forest
-----------------------------------------------------
As with any algorithm, there are advantages and disadvantages to using it. In the next two sections we'll take a look at the pros and cons of using random forest for classification and regression.

1> The random forest algorithm is not biased, since, there are multiple trees and each tree is trained on a subset of data. Basically, the random forest algorithm relies on the power of "the crowd"; therefore the overall biasedness of the algorithm is reduced.

2> This algorithm is very stable. Even if a new data point is introduced in the dataset the overall algorithm is not affected much since new data may impact one tree, but it is very hard for it to impact all the trees.

3> The random forest algorithm works well when you have both categorical and numerical features.

4> The random forest algorithm also works well when data has missing values or it has not been scaled well.

Disadvantages of using Random Forest
---------------------------------------------------------
1> A major disadvantage of random forests lies in their complexity. They required much more computational resources, owing to the large number of decision trees joined together.

2> Due to their complexity, they require much more time to train; than other comparable algorithms.

1> Using Random Forest for Regression
----------------------------------------------------------

<font color='red'><u>Problem Definition</u> : The problem here is to <b>predict the gas consumption</b> (in millions of gallons) in 48 of the US states based on petrol tax (in cents), per capita income (dollars), paved highways (in miles) and the proportion of population with the driving license. </font>

In [ ]:
# Steps to upload any dataset into your Colab NB :
# step 1 : First Download the dataset to your local PC.
#          The link for downloading our first dataset for practicing is https://drive.google.com/open?id=1CSP3e6k-04UZZoQd_6NP_wUSts4wNLIc
# step 2 : Run the below code and select the (above downloaded) dataset.

#from google.colab import files
#file = files.upload()

In [ ]:
# Import Libraries
import pandas as pd
import numpy as np

dataset = pd.read_csv('petrol_consumption.csv')
dataset.head()

Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)  \
0         9.0            3571            1976                         0.525   
1         9.0            4092            1250                         0.572   
2         9.0            3865            1586                         0.580   
3         7.5            4870            2351                         0.529   
4         8.0            4399             431                         0.544   

   Petrol_Consumption  
0                 541  
1                 524  
2                 561  
3                 414  
4                 410

In [ ]:
# Preparing the Data
# divide the data into attributes and labels

X = dataset.drop('Petrol_Consumption', axis=1)
y = dataset['Petrol_Consumption']

# dividing data into training and testing set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Training and Making Predictions
from sklearn.ensemble import RandomForestRegressor
classifier = RandomForestRegressor(n_estimators=20, random_state=42)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)


# The RandomForestRegressor class of the sklearn.ensemble library
# is used to solve regression problems via random forest.
# The most important parameter of the RandomForestRegressor class
# is the n_estimators parameter.
# This parameter defines the number of trees in the random forest.

# Evaluating the Algorithm
from sklearn import metrics

print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 51.76500000000001
Mean Squared Error: 4216.166749999999
Root Mean Squared Error: 64.93201637097064


With 20 trees, the root mean squared error is 64.93 which is greater than 10 percent of the average petrol consumption ( i.e. 10% of 576.77 = 57.67). This may indicate, among other things, that we have not used enough estimators (trees).

In [ ]:
# If the number of estimators is changed to 200, the results are as follows:

from sklearn.ensemble import RandomForestRegressor
classifier = RandomForestRegressor(n_estimators=200, random_state=42)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)


# Evaluating the Algorithm
from sklearn import metrics

print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred))
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 48.124500000000005
Mean Squared Error: 3497.802072500001
Root Mean Squared Error: 59.142219035981405


<font color='green'><b>Analysis :</b></font> With 200 trees, the root mean squared error is 59.14 which is close to 10 percent of the average petrol consumption (i.e. 10% of 576.77 = 57.67).

<font color='red'>The Student is expected to try <b> n_estimators</b> values of 220 and 250. </font>

<b>Do you observe any difference ? </b><br>
<font color='green'>If yes then please share on the whats-app group. This would help your peers in learning better. </font>

2: Using Random Forest for Classification
------------------------------------------------------------

<font color='red'><b>Problem Definition</b> : The task here is to predict whether a bank currency note is <u>authentic</u> or <u>not</u> based on four attributes i.e. variance of the image wavelet transformed image, skewness, entropy, and curtosis of the image.</font>

In [ ]:
# Steps to upload any dataset into your Colab NB :
# step 1 : First Download the dataset to your local PC.
#          The link for downloading our first dataset for practicing is https://drive.google.com/open?id=19YvsKMdlIZ_bxgJOSg4waIkVTyJ-UYx_
# step 2 : Run the below code and select the (above downloaded) dataset.

#from google.colab import files
#file = files.upload()

In [ ]:
# doing the minimum necessary imports
# more modules would be imported as and when needed

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

# reading data from CSV file.
# reading bank currency note data into pandas dataframe.
bankdata = pd.read_csv("bill_authentication.csv")

# Exploratory Data Analysis
print(bankdata.shape)
print("------------")
print(bankdata.head())

(1372, 5)
------------
   Variance  Skewness  Curtosis  Entropy  Class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0


In [ ]:
# Data Preprocessing
# Data preprocessing involves
# (1) Dividing the data into attributes and labels and
# (2) dividing the data into training and testing sets.

# To divide the data into attributes and labels, do :
X = bankdata.drop('Class', axis=1)
y = bankdata['Class']

# the final preprocessing step is to divide data into training and test sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20)


# Training and predict the Algorithm.




# Evaluating the Algorithm
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))
print(accuracy_score(y_test, y_pred))

[[147   3]
 [  0 125]]
              precision    recall  f1-score   support

           0       1.00      0.98      0.99       150
           1       0.98      1.00      0.99       125

    accuracy                           0.99       275
   macro avg       0.99      0.99      0.99       275
weighted avg       0.99      0.99      0.99       275

0.9890909090909091


<font color='green'><b>Analysis</b></font> : The accuracy achieved for by our random forest classifier with 20 trees is 99% ( which is great !!).

<font color='red'>The Student is expected to try n_estimators values of 10 and 50. </font>

<b>Do you observe any difference ?</b> <br>
<font color='green'>If yes then please share on the whats-app group. This would help your peers in learning better. </font>

Q> <font color='red'>How do I get the best value of the estimators ? </font>
**[remember : its not always best to have higher no. of estimators?**] <br>[<i><font color='green'> Becoz higher no. of estimators means more trees , which means more time.</font></i> ] <br>

In [ ]:
# follow hints to answer above Qn. of finding the best value of estimators.
# loop through values from 10 to 200, step by 10. like -> for n_estimator in range(10, 200, 10)
# use n_estimator as variable in RandomForestClassifier or  RandomForestRegressor






#-- Now please plot the error values vs. n_estimator







<hr>
<h3><font color='green'>You have coded and found the best value of n_estimators, Trust me -> You are Awesome !!</font></h3>
<hr>